packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts

```diff
@@ -193,24 +194,114 @@ export class SummarizerNode implements IRootSummarizerNode {
+	 * In case of failure, additional information is returned indicating type of failure and where it was.
+	 */
+	public validateSummary(): ValidateSummaryResult {
+		return this.validateSummaryCore(false /* parentSkipRecursion */);
+	}
+
+	/**
+	 * Validates that the in-progress summary is correct for all nodes, i.e., summarize should have run for all
+	 * non-skipped nodes.
+	 * @param parentSkipRecursion - true if the parent of this node skipped recursing the child nodes when summarizing.
+	 * In that case, the children will not have work-in-progress state.
+	 *
+	 * @returns ValidateSummaryResult which contains a boolean success indicating whether the validation was successful.
+	 * In case of failure, additional information is returned indicating type of failure and where it was.
+	 */
+	protected validateSummaryCore(parentSkipRecursion: boolean): ValidateSummaryResult {
+		if (!this.didSummarize(parentSkipRecursion)) {
+			return {
+				success: false,
+				reason: "NodeDidNotSummarize",
+				id: {
+					tag: TelemetryDataTag.CodeArtifact,
+					value: this.telemetryNodeId,
+				},
+				// These errors are usually transient and should go away when summarize is retried.
+				retryAfterSeconds: 1,
+			};
+		}
+		if (parentSkipRecursion) {
+			return { success: true };
+		}
+
+		for (const child of this.children.values()) {
+			const result = child.validateSummaryCore(this.wipSkipRecursion || parentSkipRecursion);
+			// If any child fails, return the failure.
+			if (!result.success) {
+				return result;
+			}
+		}
+		return { success: true };
+	}
+
+	private didSummarize(parentSkipRecursion: boolean): boolean {
+		assert(
+			this.wipSummaryLogger !== undefined,
+			"wipSummaryLogger should have been set in startSummary or ctor",
+		);
+		assert(this.wipReferenceSequenceNumber !== undefined, "Not tracking a summary");
+
+		// If the parentSkipRecursion is true, the parent node did not call summarize on this node. Return success.
+		if (parentSkipRecursion || this.wipLocalPaths !== undefined) {
+			return true;
+		}
+
+		/**
+		 * The absence of wip local path indicates that summarize was not called for this node. Return failure.
+		 * This can happen if:
+		 * 1. A child node was created after summarize was already called on the parent. For example, a data store
+		 * is realized (loaded) after summarize was called on it creating summarizer nodes for its DDSes. In this case,
+		 * parentSkipRecursion will be true and the if block above would handle it.
+		 * 2. A new node was created but summarize was never called on it. This can mean that the summary that is
+		 * generated may not have the data from this node. We should not continue, log and throw an error. This
+		 * will help us identify these cases and take appropriate action.
+		 *
+		 * This happens due to scenarios such as data store created during summarize. Such errors should go away when
+		 * summarize is attempted again.
+		 */
+		return false;
```

Comment:
In light of the giant comment below, I'd split this up:

```suggestion
		// If parentSkipRecursion is true, the parent node did not call summarize on this node.
		// <<<<<<<<<PUT SOMETHING HERE EXPLAINING WHY THEN IT MAKES SENSE TO RETURN TRUE>>>>>>>>
		if (parentSkipRecursion) {
			return true;
		}

		/**
		 * The absence of wip local path indicates that summarize was not called for this node. Return failure.
		 * This can happen if:
		 * 1. A child node was created after summarize was already called on the parent. For example, a data store
		 * is realized (loaded) after summarize was called on it creating summarizer nodes for its DDSes. In this case,
		 * parentSkipRecursion will be true and the if block above would handle it.
		 * 2. A new node was created but summarize was never called on it. This can mean that the summary that is
		 * generated may not have the data from this node. We should not continue, log and throw an error. This
		 * will help us identify these cases and take appropriate action.
		 *
		 * This happens due to scenarios such as data store created during summarize. Such errors should go away when
		 * summarize is attempted again.
		 */
		return (this.wipLocalPaths === undefined);
```